src/common/utilities/FormulaEvaluation.ts

```diff
@@ -0,0 +1,641 @@
+                                topStackPrecedence.precedence <= currentTokenPrecedence.precedence :
+                                topStackPrecedence.precedence < currentTokenPrecedence.precedence
+                        )) {
+                            break;
+                        }
+
+                        output[parenDepth].push(stack[parenDepth].pop() as Token);
+                    }
+
+                    stack[parenDepth].push(token);
+                }
+            }
+        }
+
+        // When there are no more tokens to read, pop any remaining tokens from the stack to the output
+        while (stack[parenDepth].length > 0) {
+            output[parenDepth].push(stack[parenDepth].pop() as Token);
+        }
+
+        // Combine all outputs
+        let finalOutput: Token[] = [];
+        for (let i = 0; i <= parenDepth; i++) {
+            finalOutput = finalOutput.concat(output[i]);
+        }
+
+        return finalOutput;
+    }
+
+    private _getFnArity(fnName: string): number {
+        switch (fnName) {
+            case "if":
+            case "substring":
+            case "replace":
+            case "replaceAll":
+            case "padStart":
+            case "padEnd":
+            case "ternary":
+                return 3;
+            case "pow":
+            case "indexOf":
+            case "lastIndexOf":
+            case "join":
+            case "startsWith":
+            case "endsWith":
+            case "split":
+            case "addDays":
+            case "addMinutes":
+            case "appendTo":
+            case "removeFrom":
+                return 2;
+            default:
+                return 1;
+        }
+    }
+
+    public buildAST(postfixTokens: Token[]): ASTNode {
+
+        // Tokens are arranged on a stack/array of node objects 
+        const stack: (Token | ASTNode | ArrayLiteralNode)[] = [];
+        for (const token of postfixTokens) {
+            if (token.type === "STRING" || token.type === "NUMBER" || token.type === "BOOLEAN" || token.type === "VARIABLE") {
+                // Strings, numbers, booleans, function names, and variable names are pushed directly to the stack
+                stack.push(token);
+            } else if (token.type === "UNARY_MINUS") {
+                // Unary minus has a single operand, we discard the minus token and push an object representing a negative number
+                const operand = stack.pop() as ASTNode;
+                const numericValue = parseFloat(operand.value as string);
+                stack.push({ type: operand.type, value: -numericValue });
+            } else if (token.type === "OPERATOR") {
+                // Operators have two operands, we pop them from the stack and push an object representing the operation
+                if (["+", "-", "*", "/", "==", "!=", "<>", ">", "<", ">=", "<=", "&&", "||", "%", "&", "|", "?", ":"].includes(token.value as string)) {
```

Comment:
make this array a const?

Reply:
This array is now `operatorTypes` - a const outside of the class.